## 1. Python Intro

In [1]:
print("Hello, World!")

Hello, World!


### Analisis

Kode print ("Hello, World!") digunakan untuk menampilkan teks ke layar.
Fungsi print() adalah perintah bawaan (built-in function) di Python yang berfungsi untuk mencetak atau menampilkan output ke konsol.

## 2. Python Syntax

In [6]:
if 5 > 2:
  print("Lima lebih besar daripada dua!")

Lima lebih besar daripada dua!


### Analisis

Kode tersebut menggunakan pernyataan if untuk memeriksa apakah nilai 5 lebih besar dari 2. Karena kondisi 5 > 2 bernilai benar (True), maka program akan menjalankan perintah di dalam blok if, yaitu menampilkan teks "Lima lebih besar daripada dua!" ke layar. Indentasi (spasi di depan perintah print) menunjukkan bahwa perintah tersebut termasuk dalam blok if. Jika kondisi bernilai salah (False), maka perintah di dalam if tidak akan dijalankan.

## 3. Python Statements

In [7]:
print("Hello World!")
print("Have a good day.")

Hello World!
Have a good day.


### Analisis

Kode tersebut menggunakan fungsi print() untuk menampilkan teks ke layar. Baris pertama menampilkan "Hello World!", dan baris kedua menampilkan "Have a good day.". Kedua perintah dijalankan secara berurutan sehingga hasilnya akan muncul di dua baris berbeda pada output.

## 4. Python Output (Text)

In [8]:
print("This will work!")
print('This will also work!')

This will work!
This will also work!


### Analisis

Kode tersebut menggunakan fungsi print() untuk menampilkan teks ke layar. Python mengizinkan penggunaan tanda kutip ganda (" ") maupun tanda kutip tunggal (' ') untuk menulis string. Baris pertama menampilkan "This will work!", dan baris kedua menampilkan 'This will also work!' ke layar.

## 5. Python Output (Numbers)

In [11]:
print(3 + 3)

6


### Analisis

Kode tersebut menggunakan fungsi print() untuk menampilkan hasil dari operasi penjumlahan 3 + 3 lalu menampilkannya ke layar.

## 6. Python Comments

In [12]:
print("Selamat pagi!") #This is a comment

Selamat pagi!


### Analisis

Kode tersebut menggunakan fungsi print() untuk menampilkan teks "Selamat pagi!" ke layar. Tanda pagar (#) digunakan untuk menulis komentar, yaitu teks yang tidak akan dieksekusi oleh program.

## 7. Python Variables

In [13]:
x = 5
y = "John"
print(x)
print(y)

5
John


Kode tersebut membuat dua variabel, yaitu x dengan nilai 5 dan y dengan nilai "John". Variabel x menyimpan data bertipe angka (integer), sedangkan y menyimpan data bertipe teks (string).

## 8. Python Variable - Names

In [21]:
myvar = "Iin"
my_var = "Ray"
_my_var = "Haru"
myVar = "AR"
MYVAR = "Langit"
myvar2 = "Ocean"

print(myvar)
print(my_var)
print(_my_var)
print(myVar)
print(MYVAR)
print(myvar2)

Iin
Ray
Haru
AR
Langit
Ocean


### Analisis

Nama variabel harus diawali huruf atau garis bawah (_), tidak boleh diawali angka, hanya boleh mengandung huruf, angka, dan garis bawah, serta bersifat case-sensitive (huruf besar dan kecil dianggap berbeda)

## 9. Python Variables - Assign Multiple Values

In [22]:
x, y, z = "Orange", "Banana", "Cherry"
print(x)
print(y)
print(z)

Orange
Banana
Cherry


### Analisis

Kode tersebut menggunakan multiple assignment, yaitu memberikan nilai ke beberapa variabel sekaligus dalam satu baris.

## 10. Python - Output Variables

In [24]:
x = 5
y = "John"
print(x, y)

5 John


### Analisis

Kode tersebut membuat dua variabel, x dengan nilai 5 dan y dengan nilai "John". Fungsi print(x, y) digunakan untuk menampilkan keduanya dalam satu baris dengan spasi di antara nilai tersebut.

## 11. Python - Global Variables

In [1]:
x = "awesome"

def myfunc():
  print("Python is " + x)

myfunc()

Python is awesome


### Analisis

Kode tersebut mendefinisikan variabel x dengan nilai "awesome" di luar fungsi (variabel global). Fungsi myfunc() kemudian mencetak teks "Python is " yang digabungkan dengan nilai dari variabel x menggunakan operator +. Karena x adalah variabel global, fungsi dapat mengaksesnya.

## 12. Python - Python Data Types

In [2]:
x = 5
print(type(x))

<class 'int'>


### Analisis

Kode tersebut mendefinisikan variabel x dengan nilai 5, kemudian menggunakan fungsi type() untuk menampilkan tipe datanya. Fungsi type() berfungsi mengetahui jenis data dari suatu nilai atau variabel.

## 13. Python Numbers

In [4]:
x = 1
y = 2.8
z = 1j

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'float'>
<class 'complex'>


Kode tersebut membuat tiga variabel dengan tipe data berbeda: x = 1 bertipe integer (int), y = 2.8 bertipe float (bilangan desimal), dan z = 1j bertipe complex (bilangan kompleks). Fungsi type() digunakan untuk menampilkan tipe data masing-masing variabel.

## 14. Python Casting

In [6]:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3
print(x)
print(y)
print(z)

1
2
3


### Analisis

Merupakan proses casting, yaitu mengubah tipe data suatu nilai menjadi tipe data lain menggunakan fungsi bawaan Python. Kode tersebut menggunakan fungsi int() untuk mengonversi nilai ke dalam tipe data integer.
- int(1) menghasilkan nilai 1,
- int(2.8) mengubah angka desimal menjadi 2 (dibulatkan ke bawah tanpa desimal),
- int("3") mengubah string berisi angka menjadi 3.

## 15. Python Strings

In [7]:
print("It's alright")
print("He is called 'Ray'")
print('He is called "Haru"')

It's alright
He is called 'Ray'
He is called "Haru"


### Analisis

Kode tersebut menggunakan fungsi print() untuk menampilkan teks yang berisi tanda kutip. Python mengizinkan penggunaan kutip tunggal (' ') dan kutip ganda (" ") secara bergantian agar tidak terjadi konflik dalam penulisan string.
- Baris pertama menggunakan kutip ganda sehingga tanda kutip tunggal di dalam teks tidak menyebabkan error.
- Baris kedua menggunakan kutip ganda di luar dan kutip tunggal di dalam teks.
- Baris ketiga sebaliknya, menggunakan kutip tunggal di luar dan kutip ganda di dalam teks.

## 16. Python - Slicing Strings

In [8]:
b = "Hello, World!"
print(b[:5])

Hello


### Analisis

Kode tersebut menggunakan string slicing untuk mengambil sebagian teks dari variabel b. Nilai b[:5] berarti mengambil karakter dari indeks ke-0 hingga sebelum indeks ke-5 (karakter ke-5 tidak disertakan). Jadi, dari string "Hello, World!", bagian yang diambil adalah "Hello".


## 17. Python - Modify Strings

In [9]:
a = "Hello, World!"
print(a.upper())

HELLO, WORLD!


### Analisis

Kode tersebut menggunakan metode upper() pada string untuk mengubah semua huruf menjadi huruf kapital.

## 18. Python - String Concatenation

In [10]:
a = "Hello"
b = "World"
c = a + b
print(c)

HelloWorld


### Analisis

Kode tersebut melakukan penggabungan string (concatenation) antara variabel a dan b menggunakan operator +. Variabel a berisi "Hello" dan b berisi "World", sehingga hasil penggabungannya menjadi "HelloWorld" tanpa spasi.

## 19. Python - Format - Strings

In [11]:
age = 19
txt = f"My name is Iin, I am {age}"
print(txt)

My name is Iin, I am 19


### Analisis

Kode tersebut menggunakan f-string (formatted string literal) untuk menyisipkan nilai variabel ke dalam teks. Variabel age bernilai 19, lalu pada string f"My name is Iin, I am {age}", tanda kurung kurawal {} digunakan untuk menempatkan nilai variabel age secara langsung di dalam kalimat.

## 20. Python - Escape Characters


In [13]:
txt = "We are the so-called \"Vikings\" from the north."
print(txt) 

We are the so-called "Vikings" from the north.


### Analisis

Kode tersebut menggunakan escape character (\") untuk menampilkan tanda kutip ganda di dalam string tanpa menyebabkan error. Tanda \ memberi tahu Python bahwa tanda kutip berikutnya adalah bagian dari teks, bukan penutup string.

## 21. Python Booleans


In [15]:
a = 200
b = 33

if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


### Analisis

Kode tersebut menggunakan percabangan if–else untuk membandingkan dua nilai. Variabel a bernilai 200 dan b bernilai 33. Kondisi b > a akan menghasilkan False, karena 33 tidak lebih besar dari 200. Oleh karena itu, program menjalankan blok else.

## 22. Python Operators


In [16]:
sum1 = 100 + 50      # 150 (100 + 50)
sum2 = sum1 + 250    # 400 (150 + 250)
sum3 = sum2 + sum2   # 800 (400 + 400)

print(sum1)
print(sum2)
print(sum3)

150
400
800


### Analisis

Kode tersebut melakukan operasi penjumlahan bertahap dengan menyimpan hasilnya ke dalam variabel baru.
- sum1 = 100 + 50 menghasilkan 150,
- sum2 = sum1 + 250 menghasilkan 400,
- sum3 = sum2 + sum2 menghasilkan 800.
Fungsi print() menampilkan ketiga hasil tersebut ke layar:

## 23. Python Arithmetic Operators

In [17]:
x = 15
y = 4

print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)
print(x ** y)
print(x // y)

19
11
60
3.75
3
50625
3


### Analisis

Kode tersebut melakukan berbagai operasi aritmatika antara variabel x = 15 dan y = 4.
Berikut penjelasan setiap operasi dan hasilnya:
- x + y → penjumlahan → 15 + 4 = 19
- x - y → pengurangan → 15 - 4 = 11
- x * y → perkalian → 15 × 4 = 60
- x / y → pembagian → 15 ÷ 4 = 3.75
- x % y → sisa bagi (modulus) → sisa dari 15 ÷ 4 = 3
- x ** y → pemangkatan → 15⁴ = 50625
- x // y → pembagian bulat (floor division) → hasil bagi tanpa desimal = 3

## 24. Python Assignment Operators

In [18]:
numbers = [1, 2, 3, 4, 5]
count = len(numbers)
if count > 3:
    print(f"List has {count} elements")

if (count := len(numbers)) > 3:
    print(f"List has {count} elements")

List has 5 elements
List has 5 elements


### Analisis

Kode tersebut menggunakan dua cara untuk memeriksa jumlah elemen dalam list numbers = [1, 2, 3, 4, 5].
- Pada bagian pertama, fungsi len(numbers) menghitung jumlah elemen (5) dan menyimpannya ke variabel count. Karena count > 3, program menampilkan "List has 5 elements".
- Pada bagian kedua, digunakan walrus operator (:=) yang memungkinkan penugasan nilai di dalam ekspresi kondisi. Di sini, count langsung diisi dengan hasil len(numbers) di dalam pernyataan if, tanpa harus dideklarasikan terpisah.

## 25. Python Comparison Operators

In [19]:
x = 5
y = 3

print(x == y)
print(x != y)
print(x > y)
print(x < y)
print(x >= y)
print(x <= y)

False
True
True
False
True
False


### Analisis

Kode tersebut menggunakan operator perbandingan untuk membandingkan nilai antara variabel x = 5 dan y = 3. Setiap pernyataan menghasilkan nilai logika True atau False sesuai kondisi berikut:
- x == y → 5 sama dengan 3 → False
- x != y → 5 tidak sama dengan 3 → True
- x > y → 5 lebih besar dari 3 → True
- x < y → 5 lebih kecil dari 3 → False
- x >= y → 5 lebih besar atau sama dengan 3 → True
- x <= y → 5 lebih kecil atau sama dengan 3 → False

## 26. Python Logical Operators

In [21]:
x = 5

print(x < 5 or x > 10)

False


### Analisis

Kode tersebut menggunakan operator logika OR untuk mengevaluasi dua kondisi: x < 5 dan x > 10. Nilai x adalah 5, sehingga kedua kondisi bernilai False. Karena operator or akan menghasilkan True jika salah satu kondisi benar, dan di sini keduanya salah, maka hasil akhirnya adalah False.

## 27. Python Identity Operators

In [22]:
x = ["apple", "banana"]
y = ["apple", "banana"]
z = x

print(x is z)
print(x is y)
print(x == y)

True
False
True


### Analisis

Kode tersebut membandingkan tiga variabel x, y, dan z menggunakan operator identitas (is) dan operator kesetaraan (==).
- x is z → True, karena z menunjuk ke objek yang sama dengan x (bukan hanya nilainya yang sama).
- x is y → False, karena meskipun isi list-nya sama, x dan y adalah dua objek berbeda di memori.
- x == y → True, karena isi (elemen) dari kedua list sama persis.

## 28. Python Membership Operators

In [24]:
text = "Hello World"

print("H" in text)
print("hello" in text)
print("z" not in text)

True
False
True


### Analisis

Kode tersebut menggunakan operator keanggotaan (in dan not in) untuk memeriksa apakah suatu karakter atau teks terdapat di dalam string text = "Hello World".
- "H" in text → True, karena huruf H ada di dalam teks.
- "hello" in text → False, karena Python membedakan huruf besar dan kecil (case-sensitive).
- "z" not in text → True, karena huruf z tidak terdapat di dalam teks.

## 29. Python Bitwise Operators

In [26]:
# The & operator compares each bit and set it to 1 if both are 1, otherwise it is set to 0
# 6 = 0110
# 3 = 0011
# --------
# 2 = 0010

print(6 & 3)

2


### Analisis

Kode tersebut menggunakan operator bitwise AND (&) yang membandingkan setiap bit dari dua bilangan biner. Setiap posisi bit akan bernilai 1 hanya jika kedua bit pada posisi tersebut sama-sama 1.

## 30. Python Operator Precedence

In [28]:
print(100 + 5 * 3)

115


### Analisis

Kode tersebut melakukan operasi aritmatika dengan aturan prioritas operator (operator precedence). Dalam Python, perkalian (*) memiliki prioritas lebih tinggi daripada penjumlahan (+), sehingga dihitung terlebih dahulu.

## 32. Python If Statement

In [29]:
a = 33
b = 200

if b > a:
  print("b is greater than a")

b is greater than a


### Analisis

Kode tersebut menggunakan percabangan if untuk membandingkan dua variabel, a = 33 dan b = 200. Kondisi b > a bernilai True karena 200 lebih besar dari 33. Maka, perintah di dalam blok if dijalankan.

## 33. Python Elif Statement

In [31]:
a = 33
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")

a and b are equal


### Analisis

Kode tersebut menggunakan struktur percabangan if–elif untuk membandingkan dua variabel, a = 33 dan b = 33.
- Kondisi pertama b > a bernilai False karena nilainya sama.
- Kondisi kedua a == b bernilai True, sehingga perintah di dalam blok elif dijalankan dan menampilkan hasil:

## 34. Python Else Statement

In [32]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")

a is greater than b


### Analisis

Kode tersebut menggunakan struktur percabangan if–elif–else untuk membandingkan nilai a = 200 dan b = 33.
- Kondisi pertama b > a bernilai False,
- Kondisi kedua a == b juga False,
- Maka program menjalankan blok else, karena kedua kondisi sebelumnya tidak terpenuhi.

## 35. Python Shorthand If


In [33]:
a = 200
b = 33

if a > b: print("a is greater than b")

a is greater than b


### Analisis

Kode tersebut menggunakan pernyataan if satu baris untuk memeriksa apakah a lebih besar dari b. Karena nilai a = 200 dan b = 33, kondisi a > b bernilai True, sehingga perintah setelah tanda titik dua dijalankan.

## 36. Python Logical Operators

In [34]:
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


### Analisis

Kode tersebut menggunakan operator logika and untuk menggabungkan dua kondisi: a > b dan c > a.
- Kondisi pertama 200 > 33 bernilai True,
- Kondisi kedua 500 > 200 juga True.
Karena keduanya bernilai benar, maka ekspresi keseluruhan bernilai True, dan program menampilkan hasil:

## 37. Python Nested If

In [35]:
x = 41

if x > 10:
  print("Above ten,")
  if x > 20:
    print("and also above 20!")
  else:
    print("but not above 20.")

Above ten,
and also above 20!


### Analisis

Kode tersebut menggunakan percabangan bersarang (nested if), yaitu pernyataan if di dalam if lain.
- Pertama, kondisi x > 10 bernilai True karena 41 > 10, maka program menampilkan "Above ten,".
- Di dalamnya ada if kedua yang memeriksa x > 20, dan hasilnya juga True karena 41 > 20, sehingga program menampilkan "and also above 20!".

## 38. Python Pass Statement

In [36]:
a = 33
b = 200

if b > a:
  pass

### Analisis

Kode tersebut menggunakan pernyataan if untuk memeriksa apakah b lebih besar dari a. Karena kondisi 200 > 33 bernilai True, blok if akan dijalankan, tetapi karena berisi perintah pass, maka tidak ada aksi yang dilakukan.

Pernyataan pass digunakan sebagai placeholder, agar Python tidak menghasilkan error ketika blok kode dibiarkan kosong. Program tidak menampilkan output apa pun.

## 40. Python Match

In [37]:
day = 4
match day:
  case 1:
    print("Monday")
  case 2:
    print("Tuesday")
  case 3:
    print("Wednesday")
  case 4:
    print("Thursday")
  case 5:
    print("Friday")
  case 6:
    print("Saturday")
  case 7:
    print("Sunday")

Thursday


### Analisis

Kode tersebut menggunakan struktur kontrol match-case, fitur baru di Python 3.10 yang mirip dengan switch-case pada bahasa lain. Nilai variabel day = 4 akan dibandingkan dengan setiap case yang ada. karena nilai day sama dengan 4, maka output yang ditampilkan adalah "Thursday".